src/controller.js

```diff
@@ -0,0 +1,601 @@
+      }
+    }
+
+    let signedRequest = aws4.sign(unsignedRequest);
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest)
+    });
+    
+    let parsedResponse = parseS3Response(response.body);
+
+    let uploadId = this.__getUploadId(parsedResponse, bucket, key);
+
+    if (response.statusCode === 200) {
+      return uploadId;
+    } else {
+      throw new Error('Could not initiate multipart upload');
+    }
+  }
+
+  /**
+   * Generate the general request data for uploading an individual part of a
+   * multipart upload.  The uploadId must be provided from a previous call to
+   * the S3 initate multipart API.  The parts argument is a list of objects
+   * describing the S3 upload parts.  They should be in the format {sha256:
+   * '...', offset: int, size: int}
+   *
+   * NOTE: It is required that the process that eventually runs this request
+   * parse the response to get the Etag for the purpose of completing the
+   * upload
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html
+   */
+  async generateMultipartRequest(opts) {
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      uploadId: Joi.string().required(),
+      parts: schemas.parts.required(),
+    }));
+
+    let {bucket, key, uploadId, parts} = opts;
+    let requests = [];
+
+    for (let num = 1 ; num <= parts.length ; num++) {
+      let part = parts[num - 1];
+
+      // This is sort of hard to encapsulate in Joi for me, so for the time
+      // being we'll leave this as an external check until I can figure out how
+      // to say that all parts until the last must be at least 5MB
+      if (part.size < 5 * 1024 * 1024 && num < parts.length) {
+        throw new Error(`Part ${num}/${parts.length} must be more than 5MB, except last`);
+      }
+
+      let signedRequest = aws4.sign({
+        service: 's3',
+        region: this.region,
+        method: 'PUT',
+        protocol: this.s3protocol,
+        hostname: this.__s3hostname(bucket),
+        path: `/${key}?partNumber=${num}&uploadId=${uploadId}`,
+        headers: {
+          'x-amz-content-sha256': part.sha256,
+          'content-length': part.size,
+        }
+      });
+
+      requests.push(this.__serializeRequest(signedRequest));
+    }
+
+    return requests;
+  }
+
+  /**
+   * This method is used to tag resources, and only after completion.  This is
+   * done to maintain parity with single part uploads and should *not* be
+   * relied upon to be atomic.  This is designed to be used in things like the
+   * cost explorer.  It is not intended to be part of the public api of this
+   * library, hence the name.  If it were possible to tag the multipart upload
+   * at object creation as it is with single part uploads, we'd do that instead
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUTtagging.html
+   */
+  async __tagObject(opts) {
```

Comment:
Yes, we want it.  For single part uploads it's atomic and part of the original PUT, so it's zero overhead.

Multipart uploads are large files to begin with, so the cost of an API call vs. the size of the file makes it pay off.  It also means that we're doing a lot of API calls, 2 as setup and teardown and 1 for each part.  An extra tagging call is nothing.

My main motivation was so that we could break down our costs in the Cost Explorer, and big files (so multipart) are the most important.

There's no reason to use aws-sdk.  The XML needed is trivially simple to generate and there's full unit tests for it.